operator/v1/types_ingress.go

```diff
@@ -585,6 +591,51 @@ type EndpointPublishingStrategy struct {
 	NodePort *NodePortStrategy `json:"nodePort,omitempty"`
 }
 
+// ClientCertificatePolicy describes the policy for client certificates.
+// +kubebuilder:validation:Enum=Required;Optional
+type ClientCertificatePolicy string
+
+const (
+	// ClientCertificatePolicyRequired indicates that a client certificate
+	// should be required.
+	ClientCertificatePolicyRequired ClientCertificatePolicy = "Required"
+
+	// ClientCertificatePolicyOptional indicates that a client certificate
+	// should be requested but not required.
+	ClientCertificatePolicyOptional ClientCertificatePolicy = "Optional"
+)
+
+// ClientTLS specifies TLS configuration to enable client-to-server
+// authentication, which can be used for mutual TLS.
+type ClientTLS struct {
+	// clientCertificatePolicy specifies whether the ingress controller
+	// requires clients to provide certificates.  This field accepts the
+	// values "Required" or "Optional".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	ClientCertificatePolicy ClientCertificatePolicy `json:"clientCertificatePolicy"`
+
+	// clientCA is a reference to a configmap containing the PEM-encoded CA
+	// certificate bundle that should be used to verify a client's
+	// certificate.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	ClientCA configv1.ConfigMapNameReference `json:"clientCA"`
+
+	// allowedSubjectPatterns specifies a list of regular expressions that
+	// should be matched against the distinguished name on a valid client
+	// certificate to filter requests.  The regular expressions must use
+	// PCRE syntax.  If this list is empty, no filtering is performed.  If
+	// the list is nonempty, then at least one pattern must match a client
+	// certificate's distinguished name or else the ingress controller
+	// rejects the certificate and denies the connection.
+	//
+	// +optional
+	AllowedSubjectPatterns []string `json:"allowedSubjectPatterns,omitempty"`
```

Comment:
which list-type is this? https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

Reply:
How should I make that determination?  I didn't find any guidance in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md or https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md.  

To devise some rules for myself, I searched for `listType=(atomic|set)` and mostly found `+listType=atomic`.  Of note, I saw `+listType=atomic` used for `spec.request`, `spec.usages`, and `spec.groups` of CertificateSigningRequestSpec; `spec.tls` of Ingress; and `spec.tokenRequests` of CSIDriver, which seem kind of sort of analogous to `allowedSubjectPatterns`, inasmuch as adding a value to, or removing a value from, any of these fields grants or revokes some kind of privilege.  Meanwhile, I only found `+listType=set` in `spec.endpoints[].addresses` of EndpointSlice (even though `spec.endpoints` is itself `+listType=atomic`?!), `status.storageVersions[].decodableVersions` of StorageVersion, `spec.volumeLifecycleModes` of CSIDriver (which is immutable, so why does `listType` matter there!?), some fields in `FlowSchema`, and `registrySources.containerRuntimeSearchRegistries` of Image.config.  

From that, I propose the following guidelines:  (1) Use `listType=set` if the convenience of being able to add values with a merge is strongly desired and there is little or no risk to keeping old values by accident (e.g., no risk of accidentally failing to revoke privilege).  (2) Otherwise use `listType=atomic` as it is more common and therefore less likely to cause surprises.  

Applying my own guidelines here, I decided to use `listType=atomic`.

Reply:
Your guidelines seem pretty good, but I'm not entirely sure on why you see a risk of forgetting to remove an old value.  If the applier that added five elements issues an update that contains four elements, I thought that the the fifth would be removed.

Reply:
> If the applier that added five elements issues an update that contains four elements, I thought that the the fifth would be removed.

If that's the case, then what's the difference between `listType=atomic` and `listType=set`?  The documentation says the following:

> If configured as `atomic`, the entire list is replaced during merge.

The documentation doesn't say that that isn't the case for `listType=set`, but in the absence of other distinguishing characteristics, I assumed it wasn't.  

The documentation does also mention this:

> The `atomic` list type is recursive.

So maybe that's the distinguishing characteristic for `atomic` vis-à-vis `set`?